src/tests/test_common/homestore_test_common.hpp

```diff
@@ -234,7 +246,11 @@ class HSTestHelper {
 
             if (!fake_restart && init_device) { init_files(s_dev_names, dev_size); }
             for (const auto& fname : s_dev_names) {
-                device_info.emplace_back(std::filesystem::canonical(fname).string(), homestore::HSDevType::Data);
+                if (device_info.empty())
```

Comment:
This shouldn't be needed as well, the data type will be deduced by iomgr and set to FAST automatcially.

Reply:
can you point out the code where iomgr will detect and change pdev type?
especially where this changed `std::map< HSDevType, std::vector< PhysicalDev* > > m_pdevs_by_type`

Reply:
```
PhysicalDev::read_and_fill_superblock
--> iomgr::DriveInterface::get_attributes
   --> iomgr::DriveInterface::get_drive_type
           // <<<< if emulate_drive_type is called before, get drive type from cache, 
           // <<<< if not called before, call detect_drive_type below
           --> iomgr::DriveInterface::detect_drive_type 
```

Reply:
@yamingk 
1 `PhysicalDev::read_and_fill_superblock` I can not find this in homestore ,can you point out the line and file of this code ? 

2 `iomgr::DriveInterface::get_drive_type` yes， I know we have this, Iomanager does have the capability to detect the device type. however,  according to the current code of master branch , we do not use this capability to determine the device type, right? the device type is always specified by the input parameters?

pls see:
https://github.com/eBay/HomeStore/blob/e9721a0910eacbe36002737dfac015d8875136aa/src/lib/device/device_manager.cpp#L64

and 
https://github.com/eBay/HomeStore/blob/e9721a0910eacbe36002737dfac015d8875136aa/src/lib/device/device_manager.cpp#L129

`dinfo.dev_type` is got from the passed in parameter and never changed by IOmanager, right?

if I miss some code, pls let me know , thanks!

Reply:
```
--> DeviceManager::format_devices
  --> DeviceManager::populate_pdev_info
     --> static bool is_hdd(const std::string& devname)
        --> iomgr::DriveInterface::get_drive_type
```

Reply:
I also saw this code, but it just detects the type, but change the type recorded in homestore, right?. 
```
uint32_t DeviceManager::populate_pdev_info(const dev_info& dinfo, const iomgr::drive_attributes& attr,
                                           const uuid_t& uuid, pdev_info_header& pinfo) {
    bool hdd = is_hdd(dinfo.dev_name);

    pinfo.pdev_id = m_cur_pdev_id++;
    pinfo.mirror_super_block = hdd ? 0x01 : 0x00;
    pinfo.max_pdev_chunks = hs_super_blk::max_chunks_in_pdev(dinfo);

    auto sb_size = hs_super_blk::total_size(dinfo);
    pinfo.data_offset = hs_super_blk::first_block_offset() + sb_size;
    pinfo.size = dinfo.dev_size - pinfo.data_offset - (hdd ? sb_size : 0);
    pinfo.dev_attr = attr;
    pinfo.system_uuid = uuid;

    return pinfo.pdev_id;
}
```

but my point is all the type of a pdev is determined by this , right?
```
        auto it = m_pdevs_by_type.find(dinfo.dev_type);
        if (it == m_pdevs_by_type.end()) {
            bool happened;
            std::tie(it, happened) = m_pdevs_by_type.insert(std::pair{dinfo.dev_type, std::vector< PhysicalDev* >{}});
        }
```

if we do not input the pdev type explicitly or modify the pdev type when start , then all the pdev  will be consider as HSDevType::Data.